modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/PrepareServiceImpl.java

```diff
@@ -238,26 +240,90 @@ public CompletableFuture<QueryPlan> prepareAsync(
                 .parameters(Commons.arrayToMap(operationContext.parameters()))
                 .build();
 
-        result = prepareAsync0(parsedResult, planningContext);
+        QueryTransactionContext txContext = operationContext.txContext();
+
+        // Validate statement
+        CompletableFuture<ValidStatement> validFut = validateStatement(parsedResult, planningContext);
+
+        // Try optimize to fast
+        CompletableFuture<FastOptimizationResult> plannedOrValidated =
+                validFut.thenApply(stmt -> tryOptimizeFast(stmt, planningContext, txContext));
+
+        result = plannedOrValidated.thenCompose(optResult -> {
+            ValidStatement stmt = optResult.stmt;
+            QueryPlan plan = optResult.plan;
+
+            // If optimize fast returned a plan, return it.
+            if (plan != null) {
+                return CompletableFuture.completedFuture(plan);
+            } else {
+                // Otherwise, continue with the regular planning.
+
+                return prepareAsync0(stmt, planningContext, txContext);
+            }
+        });
 
         return result.exceptionally(ex -> {
-                    Throwable th = ExceptionUtils.unwrapCause(ex);
+            Throwable th = ExceptionUtils.unwrapCause(ex);
 
-                    throw new CompletionException(SqlExceptionMapperUtil.mapToPublicSqlException(th));
-                }
-        );
+            throw new CompletionException(SqlExceptionMapperUtil.mapToPublicSqlException(th));
+        });
+    }
+
+    private FastOptimizationResult tryOptimizeFast(
+            ValidStatement stmt,
+            PlanningContext planningContext,
+            @Nullable QueryTransactionContext txContext
+    ) {
+        // If fast query optimization is disabled, then proceed with the regular planning.
+        if (!fastQueryOptimizationEnabled()) {
+            return new FastOptimizationResult(stmt, null);
+        }
+
+        IgniteRel fastOptRel = PlannerHelper.tryOptimizeSelectCount(planningContext.planner(), txContext, stmt.value.sqlNode());
+        if (fastOptRel == null) {
+            return new FastOptimizationResult(stmt, null);
```

Comment:
Why do you need a wrapper FastOptimizationResult?
This method do not change `stmt` reference.

Reply:
Please take a look at how explain works.